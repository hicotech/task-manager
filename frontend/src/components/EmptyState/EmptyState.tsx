import type { MouseEvent } from 'react';
import { styled } from 'styled-components';

export interface Props {
    /**
     * Unique ID for the component
     */
    id?: string;

    /**
     *  On button click callback
     */
    onRequestTasks?: (e: MouseEvent<HTMLElement>) => void;

    /**
     * Allows to pass testid string for testing purposes
     */
    'data-testid'?: string;
}

const Container = styled.div`
    display: flex;
    align-items: center;
`;

const RequestTasksButton = styled.button`
    display: inline-block;
    font-weight: 700;
    font-size: 1rem;
    line-height: 1.5rem;
    cursor: pointer;
    color: #52778b;
    background-color: transparent;
    border-radius: 100rem;
    border: 2px solid #52778b;
    padding: 0.375rem 1.25rem;

    &:hover {
        color: #ffffff;
        background-color: #52778b;
    }

    &:active {
        border: 2px solid #ffffff;
        box-shadow: 0px 0px 0px 2px #52778b;
    }
`;

const EmptyState = ({ 'data-testid': dataTestId, ...props }: Props) => {
    return (
        <Container id={props.id} data-testid={dataTestId}>
            <RequestTasksButton onClick={props.onRequestTasks}>Requests Tasks</RequestTasksButton>
        </Container>
    );
};

/** @component */
export default EmptyState;
