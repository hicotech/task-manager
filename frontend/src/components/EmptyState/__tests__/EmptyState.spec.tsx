import { fireEvent, render } from '@testing-library/react';
import type { Props as EmptyStateProps } from '../EmptyState';
import EmptyState from '../EmptyState';

describe('Empty State', () => {
    const testData: Required<EmptyStateProps> = {
        id: 'emptyState-id',
        onRequestTasks: jest.fn(),
        'data-testid': 'emptyState-testid',
    };

    it('Renders Default', () => {
        const { container, getByRole } = render(<EmptyState />);

        expect(container).toMatchSnapshot();

        expect(getByRole('button')).toBeInTheDocument();
    });

    describe('Functionality testing', () => {
        it('Call onClick callback when button is clicked', () => {
            const handleClick = testData.onRequestTasks;

            const { getByRole } = render(<EmptyState onRequestTasks={handleClick} />);

            fireEvent.click(getByRole('button'));
            expect(handleClick).toHaveBeenCalledTimes(1);
        });
    });

    /**
     * Test for onRequestTasks is skipped as it would duplicate tests from "Functionality testing" section
     */
    describe('Props testing', () => {
        it('Id', () => {
            const { getByTestId } = render(
                <EmptyState id={testData.id} data-testid={testData['data-testid']} />
            );

            expect(getByTestId(testData['data-testid'])).toHaveAttribute('id', testData.id);
        });

        it('Data-testid', () => {
            const { getByTestId } = render(<EmptyState data-testid={testData['data-testid']} />);

            expect(getByTestId(testData['data-testid'])).toBeInTheDocument();
        });
    });
});
