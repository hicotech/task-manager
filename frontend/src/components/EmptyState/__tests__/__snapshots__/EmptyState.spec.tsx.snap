// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty State Renders Default 1`] = `
.c0 {
  display: flex;
  align-items: center;
}

.c1 {
  display: inline-block;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  color: #52778b;
  background-color: transparent;
  border-radius: 100rem;
  border: 2px solid #52778b;
  padding: 0.375rem 1.25rem;
}

.c1:hover {
  color: #ffffff;
  background-color: #52778b;
}

.c1:active {
  border: 2px solid #ffffff;
  box-shadow: 0px 0px 0px 2px #52778b;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1"
    >
      Requests Tasks
    </button>
  </div>
</div>
`;
